### PR TITLE
refactor: change type of linenr_T from long to int32_t

### PR DESCRIPTION
**Pull request by dundargoc**
The size of long varies depending on architecture, in contrast to the
MAXLNUM constant which sets the maximum allowable number of lines to
2^32-1. This discrepancy may lead to hard to detect bugs, for example
https://github.com/neovim/neovim/issues/18454. Setting linenr_T to a
fix maximum size of 2^32-1 will prevent this type of errors in the
future.

Also change the variables `amount` and `amount_after` to be linenr_T
since they're referring to "the line number difference" between two
texts.

### CODE REVIEW

**Comment by dundargoc** [comment]
Do this one after char_u is removed.

File: src/nvim/api/private/helpers.c [hunk 1]
```diff
@@ -1403,7 +1403,7 @@ bool set_mark(buf_T *buf, String name, Integer line, Integer col, Error *err)
       return res;
     }
   }
-  pos_T pos = { line, (int)col, (int)col };
+  pos_T pos = { (linenr_T)line, (int)col, (int)col };
```
Comment (justinmk) [inline src/nvim/api/private/helpers.c:1406]: probably need to assert here, else coverity/PVS will complain

File: src/nvim/diff.c [hunk 2]
```diff
@@ -3068,17 +3068,17 @@ static int parse_diff_ed(char_u *line, diffhunk_T *hunk)
   }
 
   if (difftype == 'a') {
-    hunk->lnum_orig = f1 + 1;
+    hunk->lnum_orig = (linenr_T)f1 + 1;
```
Comment (justinmk) [inline src/nvim/diff.c:3071]: f1 is `long` so why is this cast necessary? oh because `long` could be 64 wide :(

Does changing `f1` to int32 avoid the casts? Else we may need to assert...

File: src/nvim/shada.c [hunk 3]
```diff
@@ -3886,8 +3886,8 @@ static ShaDaReadResult shada_read_next_item(ShaDaReadDef *const sd_reader, Shada
           {
             for (i = 0; i < unpacked_2.data.via.map.size; i++) {  // -V535
               CHECK_KEY_IS_STR(unpacked_2, "buffer list entry")
-              LONG_KEY(unpacked_2, "buffer list entry", KEY_LNUM,
-                       entry->data.buffer_list.buffers[j].pos.lnum)
+              INTEGER_KEY(unpacked_2, "buffer list entry", KEY_LNUM,
+                          entry->data.buffer_list.buffers[j].pos.lnum)
```
Comment (justinmk) [inline src/nvim/shada.c:3890]: looks like `LONG_KEY` is no longer used

File: src/nvim/ex_docmd.c [hunk 4]
```diff
@@ -4443,7 +4442,7 @@ static linenr_T get_address(exarg_T *eap, char **ptr, cmd_addr_T addr_type, int
 
     default:
       if (ascii_isdigit(*cmd)) {                // absolute line number
-        lnum = getdigits_long((char_u **)&cmd, false, 0);
+        lnum = (linenr_T)getdigits_long((char_u **)&cmd, false, 0);
```
Comment (justinmk) [inline src/nvim/ex_docmd.c:4445]: do we want `getdigits_int32` ?

File: src/nvim/api/private/helpers.c (hunk 1 above)
Comment (dundargoc) [reply src/nvim/api/private/helpers.c:1406]: assert of what exactly?

We could take care of the warnings after :)

File: src/nvim/diff.c (hunk 2 above)
Comment (dundargoc) [reply src/nvim/diff.c:3071]: Yes, but need to introduce another cast.

File: src/nvim/charset.c [hunk 5]
```diff
@@ -1394,6 +1394,14 @@ long getdigits_long(char_u **pp, bool strict, long def)
   return (long)number;
 }
 
+/// Gets a int32_t number from a string.
+///
+/// @see getdigits
+int32_t getdigits_int32(char **pp, bool strict, long def)
+{
+  return (int32_t)getdigits((char_u **)pp, strict, def);
```
Comment (justinmk) [inline src/nvim/charset.c:1402]: we would need the similar treatment here as the other `getdigits_xx`. something like..

```
  intmax_t number = getdigits(pp, strict, def);
#if SIZEOF_INTMAX_T > sizeof(int32_t)
  if (strict) {
    assert(number >= INT32_MIN && number <= INT32_MAX);
  } else if (!(number >= INT32_MIN && number <= INT32_MAX)) {
    return def;
  }
#endif
  return (int32_t)number;
```

File: src/nvim/change.c [hunk 6]
```diff
@@ -380,7 +380,7 @@ void inserted_bytes(linenr_T lnum, colnr_T col, int old, int new)
 /// Takes care of marking the buffer to be redrawn and sets the changed flag.
 void appended_lines(linenr_T lnum, long count)
```
Comment (bfredl) [inline src/nvim/change.c:381]: `count` should be `linenr_T`

File: src/nvim/change.c [hunk 7]
```diff
@@ -389,20 +389,20 @@ void appended_lines_mark(linenr_T lnum, long count)
...
                    kExtmarkUndo);
   }
-  changed_lines(lnum + 1, 0, lnum + 1, count, true);
+  changed_lines(lnum + 1, 0, lnum + 1, (linenr_T)count, true);
 }
 
 /// Deleted "count" lines at line "lnum" in the current buffer.
 /// Must be called AFTER the change and after mark_adjust().
 /// Takes care of marking the buffer to be redrawn and sets the changed flag.
 void deleted_lines(linenr_T lnum, long count)
```
Comment (bfredl) [inline src/nvim/change.c:403]: `linenr_T count`

File: src/nvim/mark.c [hunk 8]
```diff
@@ -1137,7 +1136,7 @@ static void mark_adjust_internal(linenr_T line1, linenr_T line2, long amount, lo
     posp = pp; \
     if (posp->lnum == lnum && posp->col >= mincol) \
     { \
-      posp->lnum += lnum_amount; \
+      posp->lnum += (linenr_T)lnum_amount; \
```
Comment (bfredl) [inline src/nvim/mark.c:1139]: change `mark_col_adjust` instead to have `linenr_T lnum_amount` ?

File: src/nvim/api/private/helpers.c (hunk 1 above)
Comment (dundargoc) [reply src/nvim/api/private/helpers.c:1406]: Meh, random asserts create a lot of warnings as well. I'll fix it.

File: src/nvim/ex_docmd.c [hunk 9]
```diff
@@ -4519,13 +4519,13 @@ static linenr_T get_address(exarg_T *eap, char **ptr, cmd_addr_T addr_type, int
           (void)hasFolding(lnum, NULL, &lnum);
         }
         if (i == '-') {
-          lnum -= n;
+          lnum -= (linenr_T)n;
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:4522]: `n` is `long` and may overflow here

File: src/nvim/change.c [hunk 10]
```diff
@@ -224,7 +224,7 @@ static void changed_common(linenr_T lnum, colnr_T col, linenr_T lnume, long xtra
       // values for the cursor.
       // Update the folds for this window.  Can't postpone this, because
       // a following operator might work on the whole fold: ">>dd".
-      linenr_T last = lnume + xtra - 1;  // last line after the change
+      linenr_T last = lnume + (linenr_T)xtra - 1;  // last line after the change
```
Comment (zeertzjq) [inline src/nvim/change.c:227]: Can this `xtra` be changed to `linenr_T`?

Comment (dundargoc) [reply src/nvim/change.c]: Nice catch.

File: src/nvim/ex_docmd.c (hunk 9 above)
Comment (dundargoc) [reply src/nvim/ex_docmd.c:4523]: added an assert

File: src/nvim/ex_docmd.c [hunk 11]
```diff
@@ -4519,13 +4519,14 @@ static linenr_T get_address(exarg_T *eap, char **ptr, cmd_addr_T addr_type, int
           (void)hasFolding(lnum, NULL, &lnum);
         }
         if (i == '-') {
-          lnum -= n;
+          assert(INT32_MIN <= n && n <= INT32_MAX);
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:4522]: This triggers the assert: `:-3333333333print`

Comment (dundargoc) [reply src/nvim/ex_docmd.c:4522]: I don't understand what you mean.

Comment (zeertzjq) [reply src/nvim/ex_docmd.c:4522]: Can't you just change `n` to `linenr_T`?

Comment (zeertzjq) [reply src/nvim/ex_docmd.c:4522]: > I don't understand what you mean.

Build and run Nvim, type `:-3333333333print` and press, and you'll get a SIGBART.

File: src/nvim/ex_docmd.c [hunk 12]
```diff
@@ -4498,7 +4498,7 @@ static linenr_T get_address(exarg_T *eap, char **ptr, cmd_addr_T addr_type, int
       if (!ascii_isdigit(*cmd)) {       // '+' is '+1', but '+0' is not '+1'
         n = 1;
       } else {
-        n = getdigits((char_u **)&cmd, false, MAXLNUM);
+        n = (linenr_T)getdigits((char_u **)&cmd, false, MAXLNUM);
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:4501]:
```suggestion
        n = getdigits_int32((char_u **)&cmd, false, MAXLNUM);
```

**Comment by zeertzjq** [comment]
There is still something wrong with range line number overflow checking. `:3333333333yank` should give either `E16` or `E1247`, but now it no longer gives and error.

**Comment by dundargoc** [comment]
The overflow check was borked since before, `33333333333333333333print` won't give an error but silently default to 0. Think it should work now.

File: config/CMakeLists.txt [hunk 13]
```diff
@@ -8,6 +8,7 @@ include(CheckCSourceCompiles)
 check_type_size("int" SIZEOF_INT LANGUAGE C)
 check_type_size("long" SIZEOF_LONG LANGUAGE C)
 check_type_size("intmax_t" SIZEOF_INTMAX_T LANGUAGE C)
+check_type_size("int32_t" SIZEOF_INT32_T LANGUAGE C)
```
Comment (justinmk) [inline config/CMakeLists.txt:11]: Why is this needed? The size is right there in the name, 32 bits.

File: src/nvim/charset.c [hunk 14]
```diff
@@ -1392,6 +1392,22 @@ long getdigits_long(char_u **pp, bool strict, long def)
   return (long)number;
 }
 
+/// Gets a int32_t number from a string.
+///
+/// @see getdigits
+int32_t getdigits_int32(char **pp, bool strict, long def)
+{
+  intmax_t number = getdigits((char_u **)pp, strict, def);
+#if SIZEOF_INTMAX_T > SIZEOF_INT32_T
```
Comment (justinmk) [inline src/nvim/charset.c:1401]:
```suggestion
#if SIZEOF_INTMAX_T > sizeof(int32_t)
```

**Comment by justinmk** [comment]
> There is still something wrong with range line number overflow checking. :3333333333yank or :3333333333print should give either E16 or E1247, but now it no longer gives an error.

If CI passed, then we need a test to cover that.

**Comment by dundargoc** [comment]
> If CI passed, then we need a test to cover that.

I mean, is that not the "correct" behavior? Using `33333333333333333333print` in master doesn't trigger a warning either.

Anyway, I've reverted the behavior so ` :3333333333print` gives a warning again.

**Comment by justinmk** [comment]
> Using `33333333333333333333print` in master doesn't trigger a warning either

Hmm. https://github.com/neovim/neovim/pull/11015 fixed this behavior, was there a regression, or the tests in that PR don't cover some cases?

File: src/nvim/charset.c (hunk 14 above)
Comment (dundargoc) [reply src/nvim/charset.c:1401]: I don't understand why this has to be a macro though. Using

```c
  if (SIZEOF_INTMAX_T > sizeof(int32_t)) {
    ...
  }

```

works just as well.

Comment (justinmk) [reply src/nvim/charset.c:1401]: > I don't understand why this has to be a macro though.

To avoid coverity et al warning about useless checks.

**Comment by bfredl** [comment]
LGTM. was the `3333333` issue resolved satisfiably?

**Comment by dundargoc** [comment]
Yes, in the sense that the behavior hasn't changed from before.
